design/proposed/2020-openstackserver-controller.md

```diff
@@ -0,0 +1,160 @@
+
+* Resolve OpenStack resource parameters to specific resources (e.g. image, networks, subnets, security groups, server groups)
+* Create and delete ports
+* Create and delete volumes
+* Create and delete the server
+
+It will have no dependency on Machine or OpenStackCluster. So things it will explicitly not do include:
+
+* Attaching the Bastion floating IP
+* Creating a loadbalancer member
+* Adding a port for the default cluster network if no ports are provided
+* Referencing an AvailabilityZone from the Machine object
+* Adding default tags from the OpenStackCluster object
+
+## API
+
+Based on OpenStackMachineSpec, with modifications to accomodate the above.
+
+
+```go
+type OpenStackServerSpec struct {
+	AvailabilityZone       string
+	Flavor                 string
+	Image                  ImageParam
+	SSHKeyName             string
+	Ports                  []PortOpts
+	SecurityGroups         []SecurityGroupParam
+	Trunk                  bool
+	Tags                   []string
+	ServerMetadata         []ServerMetadata
+	ConfigDrive            *bool
+	RootVolume             *RootVolume
+	AdditionalBlockDevices []AdditionalBlockDevice
+	ServerGroup            *ServerGroupParam
+	IdentityRef            *OpenStackIdentityReference
+	FloatingIPPoolRef      *corev1.TypedLocalObjectReference
+	UserDataRef            *corev1.TypedLocalObjectReference
+}
+
+type OpenStackServerStatus struct {
+	InstanceID    optional.String
+	InstanceState *InstanceState
+	Addresses     []corev1.NodeAddress
+	Resolved      *ResolvedMachineSpec
+	Resources     *MachineResources
+	Conditions    clusterv1.Conditions
+}
+```
+
+As the new API is non-trivial we should initially create it in v1alpha1.
+
+### Upgrading an existing deployment
+
+OpenStackServer would be a new API and does not affect any existing API.
+
+In the same way that OpenStackMachine currently has an 'adoption' phase where it will adopt existing resources, OpenStackServer should adopt matching resources which it would have created. On upgrade to a new version of CAPO which manages the bastion with an OpenStackServer object, I would expect the flow to be:
+
+* Cluster controller creates OpenStackServer and waits for it to report Ready
+* OpenStackServer controller looks for existing resources matching its given spec
+* OpenStackServer adopts existing resources
+* OpenStackServer reports Ready
+
+The end-user should not be required to take any action, and the upgrade process should not involve the deletion and recreation of existing OpenStack resources.
+
+### Notes
+
+* ProviderID is not present: this is a Machine property
+* UserDataRef is added. If present it must exist.
+* AvailabilityZone is added, and refers explicitly
+* It is an error for Ports to be empty. Defaulting the cluster network must be done by the controller creating the object.
```

Comment:
for ports in the OpenStackServerSpec? Do we want the same API as `OpenStackMachineSpec` ?